Closest Airports page

lhoenig opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
Fixes #[issue_no]

**Summary of Changes**
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
This PR implements the CLOSEST AIRPORTS page in the DATA section of the MCDU. It displays the four closest airports from the current position with bearing, distance and estimated time of arrival. The pilot can also enter a fifth airport (any airport) to be tracked with the same data points.

Features include
- Continuous updates, refresh rate can be adjusted to match the real plane
- Freeze page
- Magvar correction of bearing (was slightly off without the correction)
- Very simple estimation of UTC arrival time

This PR is not neccessarily finished yet, but is at a stage where it needs feedback and can be tested.
What still needs to be completely implemented is the WIND/EFOB page.
Points that need to be validated include
- Are there any and how many spaces are there between the brackets in line 4 when no custom airport has been entered? The A320dp source looks like there are two spaces there.
- Are the visuals otherwise correct?
- The UTC arrival time estimation is very simple, basically just t = s/v with v being the ground speed. Does it need to be more complex? See the attached FCOM reference pages.

**Screenshots (if necessary)**
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
![pr2_1](https://user-images.githubusercontent.com/4353030/94163867-a4ac8b00-fe88-11ea-8975-7b7cb037f45d.PNG)
![pr2_2](https://user-images.githubusercontent.com/4353030/94163869-a5452180-fe88-11ea-934c-6b145d80c787.PNG)
![pr2_3](https://user-images.githubusercontent.com/4353030/94163865-a413f480-fe88-11ea-8811-f2f27385f8e6.PNG)
![fcom_ca_1](https://user-images.githubusercontent.com/4353030/94163915-b3933d80-fe88-11ea-959d-ecd55c1106eb.PNG)
![fcom_ca_2](https://user-images.githubusercontent.com/4353030/94163924-b5f59780-fe88-11ea-9e58-864cbc3ba4a2.PNG)


**Additional context**
Sources: FCOM, http://www.a320dp.com/A320_DP/nav-flight-management/sys-14.7.6.html, https://www.pprune.org/tech-log/359562-nearest-airports-page-737-a320.html

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):
lukas_#6140

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+class CDUAirportsMonitor {
+    static ShowPage(mcdu, reset=false) {
```
```suggestion
    static ShowPage(mcdu, reset = false) {
```

lhoenig replied:
Ah yes, I'm coming from mostly Python at the moment ;)

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+    static ShowPage(mcdu, reset=false) {
+
+        // one delta t unit is about 1.5 ms it seems
+        var update_ival_ms = 1000
```
```suggestion
        const update_ival_ms = 1000
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+        }
+
+        // we want the closest 4 APs with unlimited distance
+        var max_num_ap = 4
+        var max_dist_miles = 100000
```
```suggestion
        const max_num_ap = 4
        const max_dist_miles = 100000
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+        var max_dist_miles = 100000
+
+        if (this.total_delta_t >= update_ival_ms || !this.icao1) {
+            var ap_line_batch = new SimVar.SimVarBatch("C:fs9gps:NearestAirportItemsNumber", "C:fs9gps:NearestAirportCurrentLine")
```
```suggestion
            const ap_line_batch = new SimVar.SimVarBatch("C:fs9gps:NearestAirportItemsNumber", "C:fs9gps:NearestAirportCurrentLine")
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+            this.total_delta_t = 0
+        }
+
+        let s_per_h = 3600
+        let s_per_d = s_per_h * 24
```
```suggestion
        const s_per_h = 3600
        const s_per_d = s_per_h * 24
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+        let s_per_d = s_per_h * 24
+
+        // ETA offset in seconds
+        var eta1_s = Math.round(this.dist1 / this.gs * s_per_h)
+        var eta2_s = Math.round(this.dist2 / this.gs * s_per_h)
+        var eta3_s = Math.round(this.dist3 / this.gs * s_per_h)
+        var eta4_s = Math.round(this.dist4 / this.gs * s_per_h)
+
+        // absolute ETA in seconds
+        var utc_s  = Math.floor(SimVar.GetGlobalVarValue("ZULU TIME", "seconds"))
+        var utc1_s = (utc_s + eta1_s) % s_per_d
+        var utc2_s = (utc_s + eta2_s) % s_per_d
+        var utc3_s = (utc_s + eta3_s) % s_per_d
+        var utc4_s = (utc_s + eta4_s) % s_per_d
+
+        // components for ETAs in HHMM format
+        var h1 = Math.floor(utc1_s / s_per_h) || 0
+        var h2 = Math.floor(utc2_s / s_per_h) || 0
+        var h3 = Math.floor(utc3_s / s_per_h) || 0
+        var h4 = Math.floor(utc4_s / s_per_h) || 0
+        var m1 = Math.floor(utc1_s % s_per_h / 60) || 0
+        var m2 = Math.floor(utc2_s % s_per_h / 60) || 0
+        var m3 = Math.floor(utc3_s % s_per_h / 60) || 0
+        var m4 = Math.floor(utc4_s % s_per_h / 60) || 0
```
Maybe also consider making those arrays and indexing them instead

```suggestion
        const eta1_s = Math.round(this.dist1 / this.gs * s_per_h)
        const eta2_s = Math.round(this.dist2 / this.gs * s_per_h)
        const eta3_s = Math.round(this.dist3 / this.gs * s_per_h)
        const eta4_s = Math.round(this.dist4 / this.gs * s_per_h)

        // absolute ETA in seconds
        const utc_s  = Math.floor(SimVar.GetGlobalVarValue("ZULU TIME", "seconds"))
        const utc1_s = (utc_s + eta1_s) % s_per_d
        const utc2_s = (utc_s + eta2_s) % s_per_d
        const utc3_s = (utc_s + eta3_s) % s_per_d
        const utc4_s = (utc_s + eta4_s) % s_per_d

        // components for ETAs in HHMM format
        const h1 = Math.floor(utc1_s / s_per_h) || 0
        const h2 = Math.floor(utc2_s / s_per_h) || 0
        const h3 = Math.floor(utc3_s / s_per_h) || 0
        const h4 = Math.floor(utc4_s / s_per_h) || 0
        const m1 = Math.floor(utc1_s % s_per_h / 60) || 0
        const m2 = Math.floor(utc2_s % s_per_h / 60) || 0
        const m3 = Math.floor(utc3_s % s_per_h / 60) || 0
        const m4 = Math.floor(utc4_s % s_per_h / 60) || 0
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+            SimVar.SetSimVarValue("C:fs9gps:GeoCalcLongitude2", "degree", this.user_ap.infos.long)
+            this.brng5 = Math.round(SimVar.GetSimVarValue("C:fs9gps:GeoCalcBearing", "degree") - this.magvar)
+            this.dist5 = Math.round(SimVar.GetSimVarValue("C:fs9gps:GeoCalcDistance", "nautical miles"))
+            var eta5_s = Math.round(this.dist5 / this.gs * s_per_h)
+            var utc5_s = (utc_s + eta5_s) % s_per_d
+            var h5 = Math.floor(utc5_s / s_per_h) || 0
+            var m5 = Math.floor(utc5_s % s_per_h / 60) || 0
```
```suggestion
            const eta5_s = Math.round(this.dist5 / this.gs * s_per_h)
            const utc5_s = (utc_s + eta5_s) % s_per_d
            const h5 = Math.floor(utc5_s / s_per_h) || 0
            const m5 = Math.floor(utc5_s % s_per_h / 60) || 0
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:
```diff
@@ -0,0 +1,194 @@
+        // user-selected 5th airport
+        mcdu.onLeftInput[4] = () => {
+            if (!this.user_ap && mcdu.inOut !== '' && mcdu.inOut !== mcdu.clrValue) {
+                let ap = mcdu.inOut
```
```suggestion
                const ap = mcdu.inOut
```

Benjozork reviewed (commented):
Couple of things with JS best practices, but otherwise code looks good ! Added some suggestions to correct them, but haven't 100% confirmed the variables were not modified. Might need to check in your editor.

Needs testing and confirmation.

lhoenig commented:
Changed three things:
- Inserted degree sign after bearing values like in references
- Changed "PAGE FROZEN" to "LIST FROZEN" like in references
- Applied coding style suggestions from code review

Both visual changes can be seen in the attached image.

Also, one known issue: when a custom airport is entered, for one update cycle, the distance is reported wrongly. Then, the correct distance is displayed. Not sure if this is some async issue or just reported wrongly in the Simvars.

![pr2_4](https://user-images.githubusercontent.com/4353030/94193279-72af1f00-feb0-11ea-84da-9037283b41cf.PNG)

lhoenig commented:
Added the EFOB/WIND page with functional EFOBs.
I would have some questions regarding the EFF WIND and how to integrate it into the EFOB calculation. Hit me up on Discord if you want to discuss that with me!

![pr2_5](https://user-images.githubusercontent.com/4353030/94199569-4f3ca200-feb9-11ea-8302-63be736cb640.PNG)

camilo625 commented:
Name: Camilo625 #5575
Date of testing: Current date 26/09/2020
Version of the sim: 1.8.3.0
PR Tested: #837 

Testing technique: Takeoff and check airports distance while flying
Steps you've made to test the Pull Request:
1. Takeoff 
2. Set a cruze alitude
3. Check closest airports distances
4. Manually enter an airport and check its distance

Overall Rating: Good
Conclusions of the testing: Seems to work perfectly, maybe an IRL pilot should check this to be sure its correct.

Benjozork commented:
Hello, please take note that we have started including a CHANGELOG.md file that requires every PR to add a change-log line that describes its changes. As your PR is not merged yet, we kindly ask that you add this information following the included directives in the file.

Disclaimer: This is a copy-and-pasted message.

lhoenig commented:
Changelog entry appended

tareksabet commented:
Tested it, everything works good as described.
But quick question : the EFOB is static comparing an airport 12NM from plane with another airport at 216NM and has the same EFOB plus I can't CLR the custom Airport and the only way to remove it to return back to Date and then choose the closest airport page again.

lhoenig commented:
### **Fixed for 1.9.3.0**

Hey @tareksabet, thanks for testing!

I fixed the code regarding CLR, now it is possible to delete the custom airport this way.

I still have a few questions regarding details:
- Should the user-selected airport be reset if you leave the page and come back, like it is now?
- How many spaces are there between the blue brackets in line 5 when no custom airport has been entered? The sources look like there are two spaces there.

Regarding the EFOB page, I'm not sure if I fully understand the issue. From the FCOM source I read that the page 2 (EFOB) is automatically and always frozen when you visit it. So the values may have come out of date when they were showing wrong for you.
But the EFOB should still be correct at the moment you visit it. If they are instantly wrong, can you give steps to reproduce?

Best, Lukas

wpine215 commented:
Just fixed a merge conflict with the changelog.

Coralec commented:
I'm getting the dead displays from the latest Asobo update once again with this pull.

lukecologne commented:
You need to run build.py.

NathanInnes commented on .github/CHANGELOG.md:
```diff
@@ -15,4 +15,6 @@
 10. [OVHD] Fixed Battery Indicator Colour - @nathaninnes (Nathan Innes)
+
 1. [MISC] Removed Fuel Patch from MSFS Update 1.8.3 - @nathaninnes (Nathan Innes)
+2. [CDU] Add "Closest Airports" page - @lhoenig (Lukas Hoenig)
```
Change to current month

starvingjavi commented:
Name: Javi#2557
Date of testing: 10/04/2020
Version of the sim: 1.9.3
PR Tested: #837 

Testing technique: Started on a city with a lot of close airports. Checked to see if the list updated with newer airports and their distance.
Steps you've made to test the Pull Request:
1. Start on RWY 11 at SAEZ
2. In MCDU, click on DATA and Closest airports
3. Take off
4. Fly around different airports
5. Check the list of closest airports
6. Enter the airport SAZS, check for distance being updated while flying

Overall Rating: Good
Conclusions of the testing: Works as expected from the references and from what is described in the PR
